naclbox/keysource.go

```diff
@@ -0,0 +1,219 @@
+	EncryptedKey    string
+	EphemeralPubKey string // base64 encoded public key used for authentication
+	Nonce           string // base64 encoded unique nonce
+	CreationDate    time.Time
+}
+
+// KeyFile is the file representation of a public and private NACL BOX keypair
+type KeyFile struct {
+	PublicKey, PrivateKey string
+}
+
+// EncryptedDataKey returns the encrypted data key this master key holds
+func (key *MasterKey) EncryptedDataKey() []byte {
+	return []byte(key.EncryptedKey)
+}
+
+// SetEncryptedDataKey sets the encrypted data key for this master key
+func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
+	key.EncryptedKey = string(enc)
+}
+
+// Encrypt takes a sops data key, encrypts it with NACL BOX
+// and stores the result in the EncryptedKey field
+func (key *MasterKey) Encrypt(dataKey []byte) error {
+	// NACL BOX uses two keys pairs to seal secret data: sender and recipient.
+	// When used in Sops, we don't care much about the sender keypair, but it is
+	// required for the protocol to work, so we issue an ephemeral keypair instead
+	// and store its pubkey alongside the encrypted data key, so it can later
+	// be decrypted.
+	ephemeralPubKey, ephemeralPrivKey, err := box.GenerateKey(rand.Reader)
+	if err != nil {
+		return err
+	}
+	key.EphemeralPubKey = base64.StdEncoding.EncodeToString(ephemeralPubKey[:])
+
+	// We must use a different nonce for each message you encrypt with the
+	// same key. Since the nonce here is 192 bits long, a random value
+	// provides a sufficiently small probability of repeats.
+	var nonce [24]byte
+	_, err = io.ReadFull(rand.Reader, nonce[:])
+	if err != nil {
+		return err
+	}
+	key.Nonce = base64.StdEncoding.EncodeToString(nonce[:])
+
+	var pubKey *[32]byte
+	p, err := base64.StdEncoding.DecodeString(key.PublicKey)
+	if err != nil {
+		return err
+	}
+	pubKey = new([32]byte)
+	copy(pubKey[:], p)
+	encrypted := box.Seal(nonce[:], dataKey, &nonce, pubKey, ephemeralPrivKey)
+	key.EncryptedKey = base64.StdEncoding.EncodeToString(encrypted)
+	log.WithField("PublicKey", key.PublicKey).Info("Encryption succeeded")
+	return nil
+}
+
+// EncryptIfNeeded encrypts the provided sops' data key and encrypts it if it hasn't been encrypted yet
+func (key *MasterKey) EncryptIfNeeded(dataKey []byte) error {
+	if key.EncryptedKey == "" {
+		return key.Encrypt(dataKey)
+	}
+	return nil
+}
+
+// Decrypt decrypts the EncryptedKey field with NACL BOX private key and returns the result.
+func (key *MasterKey) Decrypt() ([]byte, error) {
+	// All we know at this stage is the public key used to encrypt the data key.
+	// We need to find the corresponding private key to perform the decryption.
+	// As a convention, we require that the private key is located under
+	// $HOME/.sops/naclbox/<sha256(pubkey)>.key
+	// where <sha256(pubkey)> is the sha256 hash of the raw public key
+	h := sha256.Sum256(key.pub[:])
+	path := fmt.Sprintf("%s/.sops/naclbox/%x.key", os.Getenv("HOME"), h[:])
```

Comment:
I'd split this into several functions to make it a bit easier to understand. There seems to be a few steps going on, e.g. loading the keys, validating them (base64 decoding and such), and actually decrypting

Reply:
Please also add an environment variable that allows to relocate they key to a different directory than `$HOME`.